### PR TITLE
Fix mypy issues on generic_tools, coloring, and facility

### PR DESCRIPTION
**Pull request by nhuet**
With this PR, we have no more issues with the command

```shell
mypy discrete_optimization/coloring/ discrete_optimization/generic_tools/ discrete_optimization/facility/
```

The goal is to have `mypy discrete_optimization` running without issues and adding this as a workflow step on each push.
This is mainly adding types when needed, changing some of them that were wrong. When some generic pattern seemed bugged, we changed it on all concerned files (even outside of coloring, facility and generic_tools) in a dedicated commit.

### CODE REVIEW

**Review by g-poveda — COMMENTED** [review]
this commit modifies a lot of files and not only colouring and facility source code

File: discrete_optimization/coloring/coloring_model.py [hunk 1]
```diff
@@ -98,23 +98,37 @@ def __init__(self, graph: Graph):
...
-    def evaluate(self, variable: ColoringSolution) -> Dict[str, float]:
-        try:
-            nb_color = variable.nb_color
-        except AttributeError:
-            nb_color = None
-        if nb_color is None:
-            variable.nb_color = len(set(variable.colors))
-            variable.nb_violations = self.count_violations(variable)
+    def evaluate(self, variable: Solution) -> Dict[str, float]:
+        if not isinstance(variable, ColoringSolution):
```
[the remark this answers is not in the document]
Comment (nhuet) [reply discrete_optimization/coloring/coloring_model.py:102]: Ok, I can use a cast instead (which is supposed to be almost transparent at runtime)?

**Comment by nhuet** [comment]
> this commit modifies a lot of files and not only colouring and facility source code

I imagine you are talking about the last commit. It was indeed modifying mutations from rcpsp and knapsack because I modified `discrete_optimization.generic_tools.mutations.mutation_util.get_attribute_for_type` which was used in it. 

To improve, I put the part on mutations in another commit, just before this last big commit. I hope this is clearer.

If you want I can again split work on generic_tools, then coloring, then facility.

File: discrete_optimization/coloring/coloring_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nhuet) [reply discrete_optimization/coloring/coloring_model.py:102]: I wanted to benchmark these changes by testing greedy coloring solver on gc_1000_9 (the biggest file). But even iterating on solve and taking median give different timings from a run to another.

I used examples/coloring/benchmark_coloring_greedy_solver.py script from my branch nht/profile_coloring. Even when _i do not change the code_ at all, if i run the script with 10 iteration, i get 10 timings around 2.6'' then 10 timings around 2.3''. With the same code. Strange, isn't it ?

[the remark this answers is not in the document]
Comment (nhuet) [reply discrete_optimization/coloring/coloring_model.py:102]: image
I did do 5 iteration in shell to call the python script which runs 200 times the greedy solver. And this is the timings... I don't understand these differences between "meta-iteration". I switched off everything else on the laptop when testing...

[the remark this answers is not in the document]
Comment (nhuet) [reply discrete_optimization/coloring/coloring_model.py:102]: After all the work on mypy (ie this pull request):
image

After all the work on mypy but without the isinstance checks in evaluate():
image


Conclusion: difficult to use this to benchmark as there is a lot of variance, but still it seems that there is not a big difference with the checks. But maybe the coloring problem is still too small to detect something...

[the remark this answers is not in the document]
Comment (nhuet) [reply discrete_optimization/coloring/coloring_model.py:102]: See all this work on my branch nht/profile_coloring

[the remark this answers is not in the document]
Comment (nhuet) [reply discrete_optimization/coloring/coloring_model.py:102]: Following your offline suggestion, i did the benchmark on  rcpsp evaluate instead. Here is the result: see this [notebook](https://github.com/nhuet/discrete-optimization/blob/nht/benchmark_rcpsp_evaluate/notebooks/benchmark_rcpsp_evaluate.ipynb)

**Comment by g-poveda** [comment]
Ok Thanks for having taken into account my worries on the is instance calls ! PR good to be accepted

**Pull request merged**